java/vector/src/main/java/org/apache/arrow/vector/BaseValueVector.java

```diff
@@ -33,7 +34,13 @@
 
   public static final String MAX_ALLOCATION_SIZE_PROPERTY = "arrow.vector.max_allocation_bytes";
   public static final int MAX_ALLOCATION_SIZE = Integer.getInteger(MAX_ALLOCATION_SIZE_PROPERTY, Integer.MAX_VALUE);
-  public static final int INITIAL_VALUE_ALLOCATION = 4096;
+  /*
+   * For all fixed width vectors, the value and validity buffers are sliced from a single buffer.
+   * Similarly, for variable width vectors, the offsets and validity buffers are sliced from a
+   * single buffer. To ensure the single buffer is power-of-2 size, the initial value allocation
```

Comment:
So what exactly is the value capacity we are allocating for. If the joint buffer that will be sliced has first 512 bytes for bitmap then we are having a value count of 4096 for the validity buffer and something less for data buffer. I remember that validity and data buffer move together (alloc, realloc and then load). Is that not true now and the value capacity of each buffer will vary slightly?

Reply:
I think we should change the vector.getValueCapacity() method to now take min (validity buffer value capacity, data buffer value capacity) since these buffers no longer move together.

Initially when default starting value capacity was 4096, we allocated 512 bytes of validity buffer and let's sat 4096*4 = 16KB of data buffer for an IntVector.  Now if we stick to the same numbers for the join buffer allocation, it is evident that we will over-allocate due to power of 2 semantics since 512 + 16KB will ask for a chunk of 32KB.

It seems like in the revised scheme we are still keeping 512 bytes for validity buffer (what I see from comments), so for the above example, we would be asking 512 + 3970*4 => something more than 16KB so still doesn't help.

However, if we validity size is 3970/8 = 496 and data buffer size is 3970*4, then we will allocate 16KB chunk since the compound size is just under 16KB. From your comments, it implies that we are still keeping 512 bytes for validity buffer so I am a bit confused here.

Reply:
sorry for the confusion - I meant to say approx. 512. I've fixed the comment now (it's 504 bytes for the bitmap). Also, added a test to verify that the additional allocation is < 5%.

Regarding the other questions 
- the validity and value buffers still move together. in both alloc/realloc, we allocate a single buffer and slice into two.
- the allocated value capacity can be slightly > INITIAL_VALUE_ALLOCATION, depending on the vector type.
- the  vector.getValueCapacity() already takes a minimum of validity buffer capacity and value buffer capacity (even before my change).

Reply:
Got it. Thanks